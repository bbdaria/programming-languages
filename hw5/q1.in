(* redefine in case the code wasn't copied *)

fun coords (3, _) = DNil
  | coords (_, 3) = DNil
  | coords (x, y) = DCons((x, y), fn () => coords (x + 1, y), fn () => coords (x, y + 1));
fun pcoords (3, _) = DNil
  | pcoords (_, 3) = DNil
  | pcoords (x, y) = (
    print ("exec: (" ^ Int.toString x ^ ", " ^  Int.toString y ^ ")\n"); 
    DCons((x, y), fn () => pcoords (x, y + 1), fn () => pcoords (x + 1,y))
  );

fun next Nil = Nil
  | next (Cons(x, xf)) = xf ();
  
fun take s 0 = []
  | take Nil _ = []
  | take (Cons (x, xf)) n = x :: take (xf ()) (n - 1);

val p = pcoords (0, 0);
val s = coords (0, 0);

print "===TEST START===\n";

(* Testing toMatr function *)
toMatrix s (3, 2);
toMatrix s (0, 3);
toMatrix s (3, 1);
toMatrix s (3, 3);

(* Testing Q *)
toMatrix (Q ()) (3, 4);
toMatrix (Q ()) (3, 3);
toMatrix (Q ()) (4, 3);

(* Testing diags *)
take (diags (Q ())) 10;
diags p;
next it;
next it;
next it;
next it;
next it;
next it;
next it;
next it;
next it;