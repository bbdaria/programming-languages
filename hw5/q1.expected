val it = () : unit
val it = [[(0,0),(0,1)],[(1,0),(1,1)],[(2,0),(2,1)]] : (int * int) list list
val it = [] : (int * int) list list
val it = [[(0,0)],[(1,0)],[(2,0)]] : (int * int) list list
val it = [[(0,0),(0,1),(0,2)],[(1,0),(1,1),(1,2)],[(2,0),(2,1),(2,2)]]
  : (int * int) list list
val it =
  [[(1,1),(2,1),(3,1),(4,1)],[(1,2),(2,2),(3,2),(4,2)],
   [(1,3),(2,3),(3,3),(4,3)]] : (int * int) list list
val it = [[(1,1),(2,1),(3,1)],[(1,2),(2,2),(3,2)],[(1,3),(2,3),(3,3)]]
  : (int * int) list list
val it =
  [[(1,1),(2,1),(3,1)],[(1,2),(2,2),(3,2)],[(1,3),(2,3),(3,3)],
   [(1,4),(2,4),(3,4)]] : (int * int) list list
val it = [(1,1),(2,1),(1,2),(3,1),(2,2),(1,3),(4,1),(3,2),(2,3),(1,4)]
  : (int * int) list
val it = Cons ((0,0),fn) : (int * int) Seq
exec: (1, 0)
val it = Cons ((1,0),fn) : (int * int) Seq
exec: (0, 1)
val it = Cons ((0,1),fn) : (int * int) Seq
exec: (2, 0)
val it = Cons ((2,0),fn) : (int * int) Seq
exec: (1, 1)
val it = Cons ((1,1),fn) : (int * int) Seq
exec: (0, 2)
val it = Cons ((0,2),fn) : (int * int) Seq
exec: (2, 1)
val it = Cons ((2,1),fn) : (int * int) Seq
exec: (1, 2)
val it = Cons ((1,2),fn) : (int * int) Seq
exec: (2, 2)
val it = Cons ((2,2),fn) : (int * int) Seq
val it = Nil : (int * int) Seq

